File: room.cpp
/*
Solution to Room by David Harmeyer himself!

The inspiration for the problem was that I asked for a prime-numbered
room to study in in the library and was given room number 16. True story.

Anyway, one method of solving this problem is to binary search the base
we are looking for. If we interprit the room we are given in too large a base,
then we will get a room number that is bigger than what we are looking for.
If we interpret it in a base that is too small, we will get a base too small.

In fact, you can take advantage of the fact that the number you are given doesn't have
that many digits at all to use a solution that doesn't use this binary search,
but that is a bit more complicated and unnecissary.

The one part to be careful with when binary searching here is that you
don't want to overflow longs when calculating something with a large base.
I handle this by first checking to see if I will be way above any feasable answer
with doubles, and then using more precise 64 bit long longs (equivalent to longs in java)
to get exactly the right answer once I am sure I won't overflow.

*/

#include <iostream>

using namespace std;
typedef long long ll;

bool baseTooBig(ll given, ll target, ll base) {
  double max=2e18;
  int digits[7];
  int takeDigitsOf=given;
  for (int i=0; i<7; i++) {
    digits[i]=takeDigitsOf%10;
    takeDigitsOf/=10;
  }
  ll basePows[7];
  basePows[0]=1;
  for (int i=1; i<7; i++) {
    if (basePows[i-1]==-1||basePows[i-1]*(double)base>max)
      basePows[i]=-1;
    else
      basePows[i]=basePows[i-1]*base;
  }
  ll total=0;
  for (int i=0; i<7; i++) {
    if (total==-1||total+(double)digits[i]*(double)basePows[i]>max
        ||(digits[i]!=0&&basePows[i]==-1))
      total=-1;
    else
      total+=digits[i]*basePows[i];
  }
  if (total==-1)
    return true;
  return total>target;
}

int main() {
    int T;
    cin>>T;
    for (int tt=1; tt<=T; tt++) {
      ll given, target;
      cin>>given>>target;
      ll low=10, hi=1000000000000000000;
      while (low!=hi) {
        ll mid=(low+hi+1)/2;
        if (baseTooBig(given, target, mid))
          hi=mid-1;
        else
          low=mid;
      }
      cout<<"Hotel Visit #"<<tt<<": Base "<<low<<"\n";
    }
}

File: room.java
import java.math.BigInteger;
import java.util.Scanner;

public class room {

	public static void main(String[] args) {
		
		Scanner in = new Scanner(System.in);
		
		// read in num cases
		int T = in.nextInt();
		for (int t = 1; t <= T; t++) {
			
			// read in g and d
			int g = in.nextInt();
			BigInteger d = in.nextBigInteger();
			
			// split number into array of digits
			int[] digits = new int[7];
			for (int q = 6; q >= 0; q--) {
				digits[q] = g % 10;
				g /= 10;
			}
			
			// bin search to find answer
			long low = 10, high = Long.MAX_VALUE / 3;
			while (true) {
				
				long base = (low + high) / 2;
				
				// calculate the number
				BigInteger num = new BigInteger("" + digits[0]);
				for (int q = 1; q < 7; q++) {
					num = num.multiply(new BigInteger("" + base)).add(new BigInteger("" + digits[q]));
				}
				
				// if exceeds a long or too big
				if (num.compareTo(d) > 0) {
					high = base;
				}
				
				// if too small
				else if (num.compareTo(d) < 0) {
					low = base;
				}
				
				// if the right answer
				else {
					System.out.println("Hotel Visit #" + t + ": Base " + base);
					break;
				}
			}
		}
		
		in.close();
	}
}